Remove LINQ and string operation from IsDPoPError.

paulomorgado opened the pull request:
**What issue does this PR address?**

Remove extra allocations caused by LINQ and `string` operations in token names key generation

**Description**

**LINQ** operations are replaced with iterations removing delegate allocation and invocation.

`string` operations are replaced with span (`ReadOnlySpan<char>`) operations, removing allocations.

**Benchmarks**

```

BenchmarkDotNet v0.13.10, Windows 11 (10.0.22631.2715/23H2/2023Update/SunValley3)
13th Gen Intel Core i9-13900K, 1 CPU, 32 logical and 24 physical cores
.NET SDK 8.0.100
  [Host]     : .NET 8.0.0 (8.0.23.53103), X64 RyuJIT AVX2
  Job-UNPWZE : .NET 6.0.25 (6.0.2523.51912), X64 RyuJIT AVX2
  Job-WDOIWV : .NET 8.0.0 (8.0.23.53103), X64 RyuJIT AVX2

```
| Method                          | Runtime  | Mean       | Ratio | Gen0   | Allocated | Alloc Ratio |
|-------------------------------- |--------- |-----------:|------:|-------:|----------:|------------:|
| UsingLinqAndStringOperations    | .NET 6.0 | 69.5984 ns | 1.338 | 0.0088 |     168 B |        1.00 |
| UsingIterationAndSpanOperations | .NET 6.0 |  0.3063 ns | 0.006 |      - |         - |        0.00 |
| UsingLinqAndStringOperations    | .NET 8.0 | 52.0763 ns | 1.000 | 0.0089 |     168 B |        1.00 |
| UsingIterationAndSpanOperations | .NET 8.0 |  0.2891 ns | 0.006 |      - |         - |        0.00 |


```csharp
[HideColumns("Error", "StdDev", "Median", "RatioSD")]
[MemoryDiagnoser]
public class IsDPoPErrorBenchmarks
{
    private const string AuthorizationHeaderDPoP = "DPoP";
    private const string Error = "error";
    private const string UseDPoPNonce = "use_dpop_nonce";

    private BasicAuthenticationHeaderValue header;

    [GlobalSetup]
    public void GlobalSetup()
    {
        this.header = new BasicAuthenticationHeaderValue(AuthorizationHeaderDPoP, $"{Error}=\"{UseDPoPNonce}\"");
    }

    [Benchmark(Baseline = true)]
    public bool UsingLinqAndStringOperations()
    {
        var header = this.header;

        if (header != null && header.Parameter != null)
        {
            // WWW-Authenticate: DPoP error="use_dpop_nonce"
            var values = header.Parameter.Split(',', StringSplitOptions.RemoveEmptyEntries);
            var error = values.Select(x =>
            {
                var parts = x.Split('=', StringSplitOptions.RemoveEmptyEntries);
                if (parts.Length == 2 && parts[0] == OidcConstants.TokenResponse.Error)
                {
                    return parts[1].Trim('"');
                }
                return null;
            }).Where(x => x != null).FirstOrDefault();

            return error == OidcConstants.TokenErrors.UseDPoPNonce || error == OidcConstants.TokenErrors.InvalidDPoPProof;
        }

        return false;
    }

    [Benchmark]
    public bool UsingIterationAndSpanOperations()
    {
        var header = this.header;

        if (header.Scheme == OidcConstants.AuthenticationSchemes.AuthorizationHeaderDPoP
            && header.Parameter is not null)
        {
            // WWW-Authenticate: DPoP error="use_dpop_nonce"
            var remaining = header.Parameter.AsSpan();

            while (!remaining.IsEmpty)
            {
                ReadOnlySpan<char> parameter;

                var separatorIndex = remaining.IndexOf(',');
                if (separatorIndex == -1)
                {
                    parameter = remaining;
                    remaining = ReadOnlySpan<char>.Empty;
                }
                else
                {
                    parameter = remaining.Slice(0, separatorIndex);
                    remaining = remaining.Slice(separatorIndex + 1).Trim(' ');
                }

                if (!parameter.IsEmpty)
                {
                    var equalsIndex = parameter.IndexOf("=");
                    if (equalsIndex != -1)
                    {
                        var name = parameter.Slice(0, equalsIndex).Trim(' ');
                        var value = parameter.Slice(equalsIndex + 1).Trim(' ').Trim('"');

                        if (name.SequenceEqual(OidcConstants.TokenResponse.Error.AsSpan()))
                        {
                            return value.SequenceEqual(OidcConstants.TokenErrors.UseDPoPNonce.AsSpan())
                                || value.SequenceEqual(OidcConstants.TokenErrors.InvalidDPoPProof.AsSpan());
                        }
                    }
                }
            }
        }

        return false;
    }
}
```

josephdecock commented:
Thanks for the PR @paulomorgado. While this does improve performance in your benchmark, it comes at the cost of more verbose code and at the loss of the nice abstractions that linq provides.  Can you make the case for why we should optimize this method so much?

paulomorgado commented:
@josephdecock, I only applied known .NET performance practices.

Once you get enough experience on how LINQ To Objects que hurt the performance of your code, what you can "nice abstractions" becomes nifes in your eyes, 😄

LINQ To Objects always comes with delegate instantiations and invocations. And, most of the times, comes with closure allocations.

`string.Trim` always creates a new string, if the string is trimmed.

`string.Split` always creates a new array and a new string for each entry.

josephdecock commented:
Sure, abstraction comes at a cost. I'm just willing to pay that cost, most of the time. To quote Knuth:

> Programmers waste enormous amounts of time thinking about, or worrying about, the speed of noncritical parts of their programs, and these attempts at efficiency actually have a strong negative impact when debugging and maintenance are considered. We should forget about small efficiencies, say about 97% of the time: premature optimization is the root of all evil. Yet we should not pass up our opportunities in that critical 3%.

paulomorgado commented:
Every quote has a timestamp and a context associated. That quote was produced in a world where hardware was bought in excess to accommodate all predictable workloads. Now we live in a world where "hardware" is paid per use by the second.

Sure there was hardware paid by some time slot at the time Knuth produced that quote. But, to achieve some improvements required changing platforms/languages for parts or all the product.

Would all those optimizations be possible to make before `Span<T>`? Sure they were. Just not as easier.

The presented code shows the full algorithm. If readability is your only issue here and you think increasing the performance of that code is worth the effort, I can make another proposal.

paulomorgado commented:
Every quote has a timestamp and a context associated. That quote was produced in a world where hardware was bought in excess to accomodate 

Would all those optimizations be possible to make before `Span<T>`? Sure they were. Just not as easier.

The presented code shows the full algorithm. If readability is your only issue here and you think increasing the performance of that code is worth the effort, I can make another proposal.

paulomorgado commented:
Hi @brockallen, whether or not one would suffer from memory pressure, depends on the number of times this code would be invoked, the size of the system (could be a very small container or a Raspberry Pi), and the overall memory pressure (the drop that makes the glass overflow).

For .NET 8, the code could be written as:

```csharp
[Benchmark]
public bool UsingStringTokenizer()
{
    var header = this.header;
    Span<Range> parametersPartsRange = stackalloc Range[2];

    if (header.Scheme == OidcConstants.AuthenticationSchemes.AuthorizationHeaderDPoP
        && header.Parameter is not null)
    {
        // WWW-Authenticate: DPoP error="use_dpop_nonce"

        foreach (var parameterSegment in new StringTokenizer(header.Parameter, [ ',' ]))
        {
            var parameterSegmentSpan = parameterSegment.AsSpan();
            parameterSegmentSpan.Split(parametersPartsRange, '=');

            if (parameterSegmentSpan[parametersPartsRange[0]].SequenceEqual(OidcConstants.TokenResponse.Error.AsSpan()))
            {
                var parameterValue = parameterSegmentSpan[parametersPartsRange[1]];
                return parameterValue.SequenceEqual(OidcConstants.TokenErrors.UseDPoPNonce.AsSpan())
                    || parameterValue.SequenceEqual(OidcConstants.TokenErrors.InvalidDPoPProof.AsSpan());
            }
        }
    }

    return false;
}
```

```
BenchmarkDotNet v0.13.10, Windows 11 (10.0.22631.3235/23H2/2023Update/SunValley3)
13th Gen Intel Core i9-13900K, 1 CPU, 32 logical and 24 physical cores
.NET SDK 8.0.200
  [Host]     : .NET 8.0.2 (8.0.224.6711), X64 RyuJIT AVX2
  Job-RMBQGJ : .NET 8.0.2 (8.0.224.6711), X64 RyuJIT AVX2

Runtime=.NET 8.0  Toolchain=net8.0

| Method                          | Mean       | Ratio | Gen0   | Allocated | Alloc Ratio |
|-------------------------------- |-----------:|------:|-------:|----------:|------------:|
| UsingLinqAndStringOperations    | 50.7521 ns | 1.000 | 0.0089 |     168 B |        1.00 |
| UsingIterationAndSpanOperations |  0.2865 ns | 0.006 |      - |         - |        0.00 |
| UsingStringTokenizer            |  0.9424 ns | 0.019 |      - |         - |        0.00 |
```

It's slightly less performant, but equally non-allocating.

The only .NET 8 depedency is [ReadOnlySpan\<Char>.Split(ReadOnlySpan\<Char>, Span\<Range>, Char, StringSplitOptions)
](https://learn.microsoft.com/en-us/dotnet/api/system.memoryextensions.split), but that could easily be implemented for .NET 6 for this particular case.

Would that be acceptable?

paulomorgado commented:
This implementation works for both .NET 6 and .NET 8:

```csharp
[Benchmark]
public bool UsingStringTokenizer()
{
    var header = this.header;

    if (header.Scheme == OidcConstants.AuthenticationSchemes.AuthorizationHeaderDPoP
        && header.Parameter is not null)
    {
        // WWW-Authenticate: DPoP error="use_dpop_nonce"

        foreach (var parameterSegment in new StringTokenizer(header.Parameter, [',']))
        {
            ExtractNameValue(parameterSegment.AsSpan(), out var name, out var value);

            if (name.SequenceEqual(OidcConstants.TokenResponse.Error.AsSpan()))
            {
                return value.SequenceEqual(OidcConstants.TokenErrors.UseDPoPNonce.AsSpan())
                    || value.SequenceEqual(OidcConstants.TokenErrors.InvalidDPoPProof.AsSpan());
            }

            [MethodImpl(MethodImplOptions.AggressiveInlining)]
            static void ExtractNameValue(ReadOnlySpan<char> nameValuePair, out ReadOnlySpan<char> name, out ReadOnlySpan<char> value)
            {
                var idx = nameValuePair.IndexOf('=');
                if (idx != -1)
                {
                    name = nameValuePair.Slice(0, idx).Trim();
                    value = nameValuePair.Slice(idx + 1).Trim();
                }
                else
                {
                    name = nameValuePair.Trim();
                    value = ReadOnlySpan<char>.Empty;
                }
            }
        }
    }

    return false;
}
```

```
BenchmarkDotNet v0.13.10, Windows 11 (10.0.22631.3235/23H2/2023Update/SunValley3)
13th Gen Intel Core i9-13900K, 1 CPU, 32 logical and 24 physical cores
.NET SDK 8.0.200
  [Host]     : .NET 8.0.2 (8.0.224.6711), X64 RyuJIT AVX2
  Job-XIUBWH : .NET 6.0.27 (6.0.2724.6912), X64 RyuJIT AVX2
  Job-HDVIKV : .NET 8.0.2 (8.0.224.6711), X64 RyuJIT AVX2


| Method                          | Runtime  | Mean       | Ratio | Gen0   | Allocated | Alloc Ratio |
|-------------------------------- |--------- |-----------:|------:|-------:|----------:|------------:|
| UsingLinqAndStringOperations    | .NET 6.0 | 69.6743 ns | 1.383 | 0.0088 |     168 B |        1.00 |
| UsingIterationAndSpanOperations | .NET 6.0 |  0.2727 ns | 0.005 |      - |         - |        0.00 |
| UsingStringTokenizer            | .NET 6.0 |  1.3907 ns | 0.028 |      - |         - |        0.00 |
| UsingLinqAndStringOperations    | .NET 8.0 | 50.3893 ns | 1.000 | 0.0089 |     168 B |        1.00 |
| UsingIterationAndSpanOperations | .NET 8.0 |  0.2930 ns | 0.006 |      - |         - |        0.00 |
| UsingStringTokenizer            | .NET 8.0 |  1.8473 ns | 0.037 |      - |         - |        0.00 |
```

Maybe it's worst a conditional implementation for .NET 6 and keep .NET 8 using `[ReadOnlySpan<Char>.Split(ReadOnlySpan<Char>, Span<Range>, Char, StringSplitOptions)](https://learn.microsoft.com/en-us/dotnet/api/system.memoryextensions.split)`.

brockallen commented:
We've decided to not accept this. We will certainly use this as input for our engineers to consider reworking ourselves. Thanks.
